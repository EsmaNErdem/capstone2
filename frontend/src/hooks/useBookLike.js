import { useState, useEffect, useContext } from 'react';
import UserContext from '../auth/UserContext';

/** Custom Hook for managing user book like
 * This hook provides functionality to manage user likes on books, including retrieving initial like status and count,
 * updating the like status and count, and handling user interactions to like or unlike a book.
 * 
 * -Utilizes useContext data to send user like to database
 * -Creates state to update book like count and user book like status
 * -Use useEffect to avoid to recalculations
 * 
 * Returns user like status, book like count states and handleLike function
 */
const useBookLike = (bookId, initialLikes, bookData={}) => {
    console.debug("useBookLike"); 

    const { hasLikedBook, likeBook } = useContext(UserContext);
    const [liked, setLiked] = useState();
    const [likes, setLikes] = useState();
    const [error, setError] = useState(null);

    // Helper function to convert "undefined" or "null" strings to actual undefined
    const convertUndefined = (value) => (value === "undefined" || value === null ? undefined : value);

    // Convert undefined/null strings in the bookData object
    const convertedBookData = Object.fromEntries(
        Object.keys(bookData).map((key) => [key, convertUndefined(bookData[key])])
    );
        
    /**By using the useEffect, the liked status is only recalculated when the id avoiding unnecessary recalculations on every render.  */
    useEffect(() => {
      setLiked(hasLikedBook(bookId));
      setLikes(initialLikes)
    }, [bookId, initialLikes]);
    
    /**
     * Send user like to database
     * Updates the like count and status based on user actions.
     */ 
    const handleLikeBook = async () => {
        try {
            setError(null);
            if (liked) {
                const unlikedBookId = await likeBook(bookId);
                if (unlikedBookId === bookId) {
                    setLikes((likes) => likes - 1);
                    setLiked(false);
                }
            } else {
                const likedBookId = await likeBook(bookId, convertedBookData);
                if (likedBookId === bookId) {
                    setLikes((likes) => likes + 1);
                    setLiked(true);
                }
            }
        } catch (error) {
            setError('Error liking book.');
            console.error('Error handling book like:', error);
        }
    };
    
  return { liked, likes, error, handleLikeBook };
};

export default useBookLike;