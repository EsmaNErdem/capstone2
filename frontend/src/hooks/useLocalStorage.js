import { useState, useEffect } from "react";

/**
 * Custom Hook to save user info like token and username into localStorage
 * 
 * Creates a user state and looks in localStorage for current user info, if cannot find it, assign null
 * When user info updates, useEffect either cleans it or updates localStorage
 * 
 * Retunrs [user, setUser]
 */

const useLocalStorage = (localStorageKey, initialValue = null) => {
    let initialUserValue
    if(localStorage.getItem(localStorageKey)) {
        try {
            initialUserValue = JSON.parse(localStorage.getItem(localStorageKey))
        } catch (e) {
            console.error("useLocalStorage custom hook: problem loading localStorage data", e);
        }
    } else {
        initialUserValue = initialValue
    }

    const [user, setUser] = useState(initialUserValue)

    useEffect(() => {
        if(user === null) {
            localStorage.removeItem(localStorageKey)
        } else {
            localStorage.setItem(localStorageKey, JSON.stringify(user))
        }
    }, [localStorageKey, user])

    return [user, setUser]
}

export default useLocalStorage;