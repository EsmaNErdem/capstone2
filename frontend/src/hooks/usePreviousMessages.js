import { useState, useEffect } from "react";
import BookClubApi from "../api";

/** Custom Hook for fetching previous messages between receiver and current user
 * This hook makes API call to backend to retrieve related data and set messaegs state
 * 
 * Returns error and messages
 */
const usePreviousMessages = (receiver, currentUserName, websocket, isOpen) => {
    const [error, setError] = useState(null);
    const [messages, setMessages] = useState([]);

    /**
     * Fetches the previous messages between users from backend sends to child component after WebSocket connection is established
     */ 
    useEffect(function getPreviousMessages() {
        console.debug("usePreviousMessages hook useEffect getPreviousMessages");
    
        const getMessages = async () => {
            try{
                if (isOpen && websocket) {
                    const roomName = `${receiver},${currentUserName}`
                    const previousMessages = await BookClubApi.getRoomPreviousMessages(roomName);
                    setMessages([...previousMessages]);                 
                }
            } catch (e) {
                console.error("usePreviousMessages useEffect API call data loading error:", e);
                setError("An error occurred while fetching previous messages.");
            }
        }
        getMessages();
    }, [isOpen, receiver, websocket]);

    return { error, messages }
}

export default usePreviousMessages;