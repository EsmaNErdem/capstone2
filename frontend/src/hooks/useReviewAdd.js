import { useState, useContext } from 'react';
import UserContext from '../auth/UserContext';

/** Custom Hook for managing user review submission on books/:id route components
 * This hook provides functionality to manage user review submission and updateting bookReviews state
 * 
 * -Utilizes useContext data to send user review to database
 * -Updates book review
 * 
 * Returns error and addBookReview function
 */
const useReviewAdd = (setBookReviews, bookData={}, errorHandlingState) => {
    console.debug("useReviewAdd"); 

    const { addUserReview } = useContext(UserContext);
    const [error, setError] = useState(null);

    // add current user's book review, function is called in BookReviewDrawer
    const addBookReview = async ({ review }) => {
        try {
            console.log({book: bookData, review})
            const newReview = await addUserReview({book: bookData, review});
            setBookReviews(r => [{...newReview, id:bookData.id}, ...r])
            return newReview
        } catch (error) {
            errorHandlingState(false)
            setError("Error adding book review.")
            console.error("Error adding book review:", error);
        }
    } 
    
    return { error, addBookReview }

}

export default useReviewAdd