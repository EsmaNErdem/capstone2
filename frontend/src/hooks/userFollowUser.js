import { useState, useEffect, useContext } from 'react';
import UserContext from '../auth/UserContext';

/** Custom Hook for managing user follow
 * This hook provides functionality to manage user follow on user profile page, retrieves initial follow status,
 * updating the follow status and user followers list, and handling user follow data to database
 * 
 * -Utilizes useContext data to send user like to database
 * -Creates state to update user follow status
 * -Use useEffect to avoid to recalculations
 * 
 * Returns user follow status, related error nand handleFollowUser function
 */
const useFollowUser = (userToBeFollowed, setUserFollow, currUserProfile, userCard, userToBeFollowedImg=null) => {
    console.debug("useFollowUser"); 
    const { hasFollowing, followUser } = useContext(UserContext);
    
    const [followed, setFollowed] = useState()
    const [error, setError] = useState(null);
    
    /**By using the useEffect, the followed status is only recalculated when username changes  */
    useEffect(() => {
        console.debug("Follow user useEffect, userToBeFollowed=", userToBeFollowed)

        setFollowed(hasFollowing(userToBeFollowed))
    }, [userToBeFollowed])
    
    /** 
     * Sends user follow and unfollow data to database
     * -Updates users followers list if user on profile view is not current user and handleFollow is called from profile follow button
     * -Updates users following list if user on profile view is current user and handleFollow is called from  user profile card follow button
    */
    const handleFollowUser = async () =>{
        try {
            setError(null);
            if (followed) {
                const unfollowedUsername = await followUser(userToBeFollowed);
                if (!currUserProfile && !userCard) {
                    // Current user is not on the profile, and the unfollow button is not from the user profile card
                    setUserFollow(followers => followers.filter(follower => follower.followedBy !== unfollowedUsername));
                } else if (currUserProfile && userCard) {
                    // Current user is on the profile, and the unfollow button is from the user profile card
                    setUserFollow(followings => followings.filter(following => following.following !== userToBeFollowed));
                } 

                setFollowed(false) 
            } else {
                const followingUser = await followUser(userToBeFollowed);
                if (!currUserProfile && !userCard) {
                    // Current user is not on the profile, and the follow button is not from the user profile card
                    setUserFollow(followers => [...followers, followingUser]);
                } else if (currUserProfile && userCard) {
                    // Current user is on the profile, and the follow button is from the user profile card
                    setUserFollow(followings => [...followings, { following: userToBeFollowed, userImg: userToBeFollowedImg }]);
                }                  

                setFollowed(true) 
            }
        } catch (error) {
            setError(`Error following user: ${error}`)
            console.error("Error handling user follow:", error);
        }
    }
    
    return { followed, error, handleFollowUser }
}

export default useFollowUser;