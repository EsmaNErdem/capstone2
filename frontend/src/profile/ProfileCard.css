.ProfileCard {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: stretch;
    padding: 0% 10%;
}

.ProfileCard-full {
    display: flex;
    flex-direction: row;
    justify-content:flex-start;
    align-items: stretch;
    padding: 0% 10%;
}

.profile-info {
    display: flex;
    flex-direction: row;
    justify-content:center;
    align-items: center;
}
