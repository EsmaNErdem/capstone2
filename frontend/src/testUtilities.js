import React from "react";
import UserContext from "./auth/UserContext";

const demoUser = {
  username: "testuser",
  firstNsme: "testfirst",
  lastName: "testlast",
  email: "test@test.net",
  img: null,
};

const UserProvider =
    ({ children, currentUser = demoUser, hasLikedReview = () => false, hasLikedBook = () => false, isUserReview = () => false, hasFollowing = () => false, setCurrentUser = u => u }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing, setCurrentUser, isUserReview }}>
      {children}
    </UserContext.Provider>
);

const CurrUserProvider =
    ({ children, currentUser = demoUser, hasLikedReview = () => false, hasLikedBook = () => false, isUserReview = () => true, hasFollowing = () => false, setCurrentUser = u => u }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing, setCurrentUser, isUserReview }}>
      {children}
    </UserContext.Provider>
);

const NonUserProvider =
    ({ children, currentUser = null, hasLikedReview = () => false, hasLikedBook = () => false, hasFollowing = () => false }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing,}}>
      {children}
    </UserContext.Provider>
);

export { UserProvider, CurrUserProvider, NonUserProvider };
