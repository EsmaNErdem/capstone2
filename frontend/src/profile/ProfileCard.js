import React, { useContext } from "react";
import { Link } from 'react-router-dom';
import Alert from "../utilities/Alert";
import UserContext from '../auth/UserContext';
import useFollowUser from "../hooks/userFollowUser";
import { Avatar, IconButton } from '@mui/material';
import "./ProfileCard.css"

/**
 * Display user profile information with follow/unfollow button
 * 
 * - Used in the user's followers and following lists in the Profile component
 */
const ProfileCard = ({ username, userImg, setUserFollow, currUserProfile }) => {
    console.debug("ProfileCard");

    const { currentUser } = useContext(UserContext);
    const { followed, error: followUser, handleFollowUser } = useFollowUser(username, setUserFollow, currUserProfile, true, userImg)

    // Check if profile card if current user's 
    const profileCardOwner = currentUser.username === username;

    return (
      <div className={profileCardOwner ? "ProfileCard-full" :  "ProfileCard"}>
        {followUser ? <Alert type="danger" messages={[followUser]} />: null}
        <div className="profile-info">
            <Link to={`/profile/${username}`} className="profile-link">
                <Avatar alt={username} src={userImg} sx={{ width: 50, height: 50, marginRight: 2, marginTop: 2 }} />
            </Link>
            <Link to={`/profile/${username}`} className="profile-link">
                <span className="profile-username">{username}</span>
            </Link>
        </div>
        {!profileCardOwner &&
        <IconButton onClick={() => handleFollowUser()} color={followed ? 'error' : 'default'} className="profile-follow-button">
            {followed ? "Unfollow" : "+ Follow"}
        </IconButton>}
      </div>
    );
  };
  
  export default ProfileCard;