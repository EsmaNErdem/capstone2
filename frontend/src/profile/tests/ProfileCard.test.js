import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { UserProvider } from "../../testUtilities";
import { render, fireEvent, act, waitFor } from '@testing-library/react';
import ProfileCard from '../ProfileCard';
import * as useFollowModule from '../../hooks/userFollowUser';

const user = {
    username: "Test2",
    userImg: "UserImg1",
}

beforeEach(() => {
    handleUserFollowMock = jest.fn();
  
    jest.spyOn(useFollowModule, 'default').mockImplementation(() => ({
      followed: false,
      error: null,
      handleFollowUser: handleUserFollowMock,
    }));
  });

afterEach(() => {
    jest.clearAllMocks();
});

test('renders without crashing', async () => {  
    await act(async () => {
      render(
        <MemoryRouter>
          <UserProvider>
            <ProfileCard />
          </UserProvider>
        </MemoryRouter>
      );
    });
  
});

test('it renders and matches with snapshot', async () => {  
    await act(async () => {
      ({ asFragment, getByTestId } = render(
        <MemoryRouter>
          <UserProvider>
            <ProfileCard />
          </UserProvider>
        </MemoryRouter>
      ));
    });
  
    expect(asFragment()).toMatchSnapshot();
});

test('renders mock data', async () => {    
    await act(async () => {
        ({ getByText, getByTestId, queryByText } = render(
        <MemoryRouter>
          <UserProvider>
            <ProfileCard 
                username={user.username}
                userImg={user.userImg}
            />
          </UserProvider>
        </MemoryRouter>
      ));
    });
    
    expect(getByText("Test2")).toBeInTheDocument()
    expect(getByText("+ Follow")).toBeInTheDocument()
});

test('renders mock follow', async () => {
    // BookClubApi.getUser.mockImplementationOnce(() => (mockUser));
    // BookClubApi.followUser.mockImplementationOnce(() => (mockFollow));
    
    await act(async () => {
      ({ getByText } = render(
          <MemoryRouter>
          <UserProvider>
            <ProfileCard 
                username={user.username}
                userImg={user.userImg}
            />
          </UserProvider>
        </MemoryRouter>
      ));
    });

    const followUser = getByText("+ Follow")
    fireEvent.click(followUser);
    
    await waitFor(() => {
        expect(handleUserFollowMock).toHaveBeenCalled();
    });

});