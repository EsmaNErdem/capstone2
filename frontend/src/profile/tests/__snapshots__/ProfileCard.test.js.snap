// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders and matches with snapshot 1`] = `
<DocumentFragment>
  <div
    class="ProfileCard"
  >
    <div
      class="profile-info"
    >
      <a
        class="profile-link"
        href="/profile/undefined"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-6vry9o-MuiAvatar-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
            data-testid="PersonIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
            />
          </svg>
        </div>
      </a>
      <a
        class="profile-link"
        href="/profile/undefined"
      >
        <span
          class="profile-username"
        />
      </a>
    </div>
    <button
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium profile-follow-button css-78trlr-MuiButtonBase-root-MuiIconButton-root"
      tabindex="0"
      type="button"
    >
      + Follow
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;
